### ISSUE TITLE
Existing Courses not upgrading progress bar in reporting

### ISSUE THREAD

**Issue by nrherron92** [gocodebox/lifterlms#1992]
### Reproduction Steps

+ HS-186321
+ On both live and staging, existing courses don't accurately update in reporting. 
+ Previously existing lessons are also not loading for edit (the user says he doesn't have this issue on the live site) I think this could be related though.
+ If you complete a lessons and then go into _LifterLMS->Status->Tools and Utilities_ to clear the reporting cache the update does not change. All of the courses remain at 0%. You can see that they are completed if you go into the student's individual course progress.
+ When creating a new course (on staging) I was able to go through it and have the progress update after clearing the reporting cache.
+ He had other lesson loading speed that this may be related to. I recommended object caching and CDN offloading for that. If they update this could it help their reporting content too?

### Expected Behavior

+ Reporting updates when reporting cache is cleared.

### Actual Behavior

+ No changes up progress is recorded.

### Error Messages / Logs

+ Include any relevant error messages or log files

```
[11-Feb-2022 18:30:10 UTC] PHP Fatal error:  Uncaught Error: Call to a member function get_sections() on null in /home/sybitkkz/public_html/staging/wp-content/plugins/lifterlms/includes/models/model.llms.section.php:243
Stack trace:
#0 /home/sybitkkz/public_html/staging/wp-content/plugins/lifterlms/includes/models/model.llms.section.php(179): LLMS_Section->get_siblings('ids')
#1 /home/sybitkkz/public_html/staging/wp-content/plugins/lifterlms/includes/admin/post-types/meta-boxes/class.llms.meta.box.course.builder.php(160): LLMS_Section->get_previous()
#2 /home/sybitkkz/public_html/staging/wp-content/plugins/lifterlms/includes/admin/post-types/meta-boxes/class.llms.meta.box.course.builder.php(131): LLMS_Metabox_Course_Builder->output_section(Object(LLMS_Section), 'previous')
#3 /home/sybitkkz/public_html/staging/wp-admin/includes/template.php(1395): LLMS_Metabox_Course_Builder->output(Object(WP_Post), Array)
#4 /home/sybitkkz/public_html/staging/wp-admin/includes/post.php(2296): do_meta_boxes(Object(WP_Screen), 'side', Object(WP_Post))
#5 /home/sybitkkz/p in /home/sybitkkz/public_html/staging/wp-content/plugins/lifterlms/includes/models/model.llms.section.php on line 243



```

### System and Environment Information

<details>
<summary>System Report</summary>

```
Wordpress
-------------------------------------------

Home Url: [removed]
Site Url: [removed]
Login Url: [removed]/wp-login.php
Version: 5.9
Debug Mode: Yes
Debug Log: Yes
Debug Display: No
Locale: en_US
Multisite: No
Page For Posts: Not Set
Page On Front: Home 3 (#10321) [[removed]/]
Permalink Structure: /%postname%/
Show On Front: page
Wp Cron: Yes


Settings
-------------------------------------------

Version: 5.8.0
Db Version: 5.8.0
Course Catalog: Course Catalog (#15) [[removed]/courses/]
Membership Catalog: Membership Catalog (#16) [[removed]/memberships/]
Student Dashboard: My Account (#8516) [[removed]/my-account/]
Checkout Page: Checkout (#6332) [[removed]/checkout/]
Course Catalog Per Page: 9
Course Catalog Sorting: date,DESC
Membership Catalog Per Page: 9
Membership Catalog Sorting: menu_order,ASC
Site Membership: Not Set
Courses Endpoint: 
Edit Endpoint: 
Lost Password Endpoint: change-password
Vouchers Endpoint: 
Autogenerate Username: no
Password Strength Meter: no
Minimum Password Strength: 
Terms Required: yes
Terms Page: Terms & Conditions (#5622) [[removed]/terms-conditions/]
Checkout Names: 
Checkout Address: 
Checkout Phone: 
Checkout Email Confirmation: no
Open Registration: yes
Registration Names: 
Registration Address: 
Registration Phone: 
Registration Voucher: 
Registration Email Confirmation: no
Account Names: 
Account Address: 
Account Phone: 
Account Email Confirmation: no
Confirmation Endpoint: confirm-payment
Force Ssl Checkout: no
Country: US
Currency: USD
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Decimals: 2
Trim Zero Decimals: no
Recurring Payments: no
Email From Address: [removed]
Email From Name: [removed]
Email Footer Text: 
Email Header Image: 
Cert Bg Width: 800
Cert Bg Height: 616
Cert Legacy Compat: no


Constants
-------------------------------------------

LLMS_REMOVE_ALL_DATA: undefined
LLMS_REST_DISABLE: undefined
LLMS_SITE_FEATURE_RECURRING_PAYMENTS: undefined
LLMS_SITE_IS_CLONE: undefined


Gateways
-------------------------------------------

Stripe: Disabled
Stripe Test Mode: Disabled
Stripe Logging: 
Stripe Order: 1
Manual: Disabled
Manual Logging: 
Manual Order: 1


Server
-------------------------------------------

Mysql Version: 5.5.5
Php Curl: Yes
Php Default Timezone: UTC
Php Fsockopen: Yes
Php Max Input Vars: 2000
Php Max Upload Size: 1 GB
Php Memory Limit: 1024M
Php Post Max Size: 1024M
Php Soap: Yes
Php Suhosin: No
Php Time Limt: 1500
Php Version: 7.4.27
Software: LiteSpeed
Wp Memory Limit: 40M


Browser
-------------------------------------------

HTTP USER AGENT: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.80 Safari/537.36


Theme
-------------------------------------------

Name: Twenty Nineteen
Version: 2.2
Themeuri: https://wordpress.org/themes/twentynineteen/
Authoruri: https://wordpress.org/
Template: 
Child Theme: No
Llms Support: No


Plugins
-------------------------------------------

LifterLMS: 5.8.0
LifterLMS Advanced Quizzes: 2.0.0
LifterLMS Assignments: 1.2.0
LifterLMS Social Learning: 1.4.0
LifterLMS Stripe Payment Gateway: 5.4.0
User Switching: 1.5.8


Integrations
-------------------------------------------

BbPress: No
BuddyPress: No
LifterLMS Social Learning: No


Template Overrides
-------------------------------------------




```

</details>

This issue has be recreated:
+ [ ] Locally
+ [x] On a staging site
+ [x] On a production website
+ [x] With only LifterLMS and a default theme

### Browser, Device, and Operating System Information

+ Browser name and version
+ Operating System name and version
+ Device name and version (if applicable)

**Comment by eri-trabiccolo** [comment]
@nrherron92 
The issue is limited to one "old" course right?
Looks like the issue is (at least) that that course has all the lessons as children of only one section BUT this section's course saved id is not the correct one.
Basically (@thomasplevy), on the staging website, I printed the value of that section `_llms_parent_course` meta, and it's an array of like 20 values (with a clear repetition):
```
(
    [0] => 4405
    [1] => 6754
    [2] => 6770
    [3] => 6985
    [4] => 6754
    [5] => 6770
    [6] => 6985
    [7] => 7597
    [8] => 5818
    [9] => 7223
    [10] => 4405
    [11] => 6754
    [12] => 6770
    [13] => 6985
    [14] => 6754
    [15] => 6770
    [16] => 6985
    [17] => 7597
    [18] => 5818
    [19] => 7223
)
```
this is very weird and I have to investigate on how it could've happened...
The correct one is `7223`, but when getting the post meta as single value (like we do in our code base) `4405` is retrieved.
Have to go now but later will check if forcing that value fixes both the issues as I think.

**Comment by eri-trabiccolo** [comment]
@nrherron92 
About the old lessons edit issue:
The issue is the one above.
I'm sorry I have no idea on how that could have been happened to be honest.
There's a way to fix it but requires custom code (2 PHP lines of code actually).

About the student progress issue. I would like to know a couple of students that had some progresses... I keep on opening student reports of people who didn't complete any lesson (except for one id=33, for which I managed to refresh the overall progress via code, and id=13, for whom I completed a lesson, and cleared the cached data).

**Comment by eri-trabiccolo** [comment]
@thomasplevy 
so the issue as said above, that the course has one section and this section's parent course is an array of like 20 courses (none of them exists, at least not anymore). I don't know how this could've happened, I cannot reproduce it, I cannot find anything in our code that would produce that .
That said, progress not working issue is related to this as well, because as soon as a student is enrolled in a course we set their course progress to 0. Then this value is updated basically only when a lesson completion status changes:
https://github.com/gocodebox/lifterlms/blob/dev/includes/models/model.llms.student.php#L1856
And the overall progress is based on that cached value (our tool to clear the student progress only clears the **overall** progress/grade).
but the lessons were completed when the `parent_course` of the lesson's section was messed up:
https://github.com/gocodebox/lifterlms/blob/dev/includes/models/model.llms.student.php#L1820

Not sure what we can do here, that db is messed up, might be due to some production <-> staging db merging back and forth.

To summarize, the issue is this one:
https://github.com/gocodebox/lifterlms/issues/1992#issuecomment-1044381747

once that is fixed, either manually on the db, or with a simple php code:
```
delete_post_meta( {section_id}, '_llms_parent_course' );
update_post_meta( {section_id}, '_llms_parent_course', {course_id} );
```
old lessons back to be editable again, and the PHP error reported by Natalie, of course, disappears.
This itself won't fix the progress reporting issue, although as soon as a student completes/marks incomplete a lesson, its course progress gets updated and the overall progress (with a cache clearance) will show up fine.

**Comment by thomasplevy** [comment]
@eri-trabiccolo 

Since we can't exactly explain it do you think the best course of action is to break our rules and get the users permission to add this code (via a snippets plugin) and run it, then remove it?

The cached progress can be manually updated by marking a lesson incomplete and recomplete it (via the reporting screens on the admin panel). Of course that's not really viable if we have a large number of students that are messed up in this way.

Do you think we should be building a tool to facilitate this?

What do you think?

**Comment by eri-trabiccolo** [comment]
@thomasplevy yeah we can break our rules.

About the tool, not sure, maybe yes. It might be massive though, ok we'll use a bg process mmm ok let's do it.

**Comment by thomasplevy** [comment]
I don't want to spend time on it if it's something taht we can't figure out how it happened.... If you and @nrherron92 think you can fix this with a PHP snippet let's just do that and move on.

**Comment by eri-trabiccolo** [comment]
Ok.
Ad I (not explicitly) said the db shows some signs of merging/migration (posts guid show different domains), so I think somewhat that might be the cause.

**Comment by eri-trabiccolo** [comment]
@nrherron92 
so as thomas said:
> can we get the users permission to add this code (via a snippets plugin) and run it, then remove it?

Once we run the code, as said before, the user will be able to edit the old lessons THEN:
a) The cached progress can be manually updated by marking a lesson incomplete and recomplete it (via the reporting screens on the admin panel).
[as far as I can see in the reporting - on the staging website - they have less than 50 students (including the test ones) for their main course - other course seem to be test courses as well], and most of them never completed any lesson at all, so for them it should not even needed.
b) Once this is done (single course progress cache updated), they can clean the (overall) progress/grade cache through the existing tool in Status -> Tools.

What do you think?

**Comment by nrherron92** [comment]
@eri-trabiccolo I'll let him know and see what the user thinks!

**Comment by eri-trabiccolo** [comment]
@nrherron92 any update on this?

**Comment by nrherron92** [comment]
@eri-trabiccolo nope the user hasn't gotten back to us

**Comment by eri-trabiccolo** [comment]
I'm closing this, let's reopen it when and if the user will reply.
